### PR TITLE
tests: Add an example to show how to interop Dafny and C#

### PR DESCRIPTION
**Pull request by cpitclaudel**
Here's the demo I gave recently, repackaged as a test.  I haven't added it to the CI yet (I'm missing the `lit` header), but the rest is ready for review.  I wrote a bit of documentation in the C# and Dafny files, so I'd appreciate some proofreading ^^

CCing @RustanLeino since you asked about how nice the proofs looked.

### CODE REVIEW

File: Test/tutorial/Simple_compiler/Compiler.dfy [hunk 1]
```diff
@@ -0,0 +1,498 @@
+// RUN: %dafny /compile:0 /spillTargetCode:3 "%s"
+// RUN: dotnet run -- example_input.calc > "%t"
```
Comment (fabiomadge) [inline Test/tutorial/Simple_compiler/Compiler.dfy:2]:
```suggestion
// RUN: dotnet run --project %S/SimpleCompiler.csproj -- %S/example_input.calc > "%t"
```

Comment (cpitclaudel) [reply Test/tutorial/Simple_compiler/Compiler.dfy]: Thanks!

**Comment by cpitclaudel** [comment]
Looks like the MacOS tests are failing as well; not sure if this is related.

```
Error: /Users/runner/.dotnet/sdk/6.0.101/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.targets(550,5): error NETSDK1177: Failed to sign apphost with error code 0: /Users/runner/work/dafny/dafny/dafny/Source/DafnyServer/obj/Release/net6.0/osx-x64/apphost: No such file or directory [/Users/runner/work/dafny/dafny/dafny/Source/DafnyServer/DafnyServer.csproj]
```

File: Test/tutorial/Simple_compiler/Compiler.dfy [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+/// Our Dafny ASTs are defined in terms of Dafny ``datatype`` values, but the AST produced using ANTLR on the C# side is defined through a class hierarchy (see ``Main.cs``).  To connect the two we need a translation function.  Here we write it in Dafny, which allows us to depend on fewer specifics of the compilation process.
+///
+/// To write the translator in Dafny we need a Dafny model of the C# class hierarchy.  Below we use traits, annotated with `{:extern}` to specify their C# names and `{:compile false}` to indicate that they should not be compiled: they act as shims that we can write the Dafny code against.  Then, once compiled to C#, our Dafny code can be linked against the real C# class hierarchy.
+///
+/// The model does not have to be completely faithful: for example we won't capture the fact that Expression and Statement share a base class on the C# side.
+///
+/// There are a few interesting difficulties: native C# types, nested types, enums:
+///
+/// 1. The C# side uses native C# types like `List<T>`.
+/// 2. C# types are commonly nested, but Dafny doesn't support nesting types.
```
Comment (keyboardDrummer) [inline Test/tutorial/Simple_compiler/Compiler.dfy:261]: Could you link to https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/nested-types to clarify what a nested type is?

File: Test/tutorial/Simple_compiler/Compiler.dfy [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+/// Connecting C# to Dafny
+/// ----------------------
+///
+/// Our Dafny ASTs are defined in terms of Dafny ``datatype`` values, but the AST produced using ANTLR on the C# side is defined through a class hierarchy (see ``Main.cs``).  To connect the two we need a translation function.  Here we write it in Dafny, which allows us to depend on fewer specifics of the compilation process.
+///
+/// To write the translator in Dafny we need a Dafny model of the C# class hierarchy.  Below we use traits, annotated with `{:extern}` to specify their C# names and `{:compile false}` to indicate that they should not be compiled: they act as shims that we can write the Dafny code against.  Then, once compiled to C#, our Dafny code can be linked against the real C# class hierarchy.
+///
+/// The model does not have to be completely faithful: for example we won't capture the fact that Expression and Statement share a base class on the C# side.
+///
+/// There are a few interesting difficulties: native C# types, nested types, enums:
```
Comment (keyboardDrummer) [inline Test/tutorial/Simple_compiler/Compiler.dfy:258]: It wasn't clear to me whether the following sections discuss how to resolve these three problems. Could you make that more clear?

File: Test/tutorial/Simple_compiler/Compiler.dfy [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+/// 2. C# types are commonly nested, but Dafny doesn't support nesting types.
+/// 3. Dafny datatypes don't compile to C# enums, so we cannot use them to model those.
+
+/// Modeling C# types in Dafny
+/// ~~~~~~~~~~~~~~~~~~~~~~~~~~
+///
+/// Native types
+/// ^^^^^^^^^^^^
+///
+/// In this example the only native C# type that we use is `List<T>`.  We add a model of the no-arguments `List<T>` constructor and of the `Add` method to build new lists.  For iterating over existing lists (passed from C#) the ``Aggregate`` method would be enough, but to make this example more informative we define a custom method `FoldL` instead, which is implemented in ``Main.cs``.
```
Comment (keyboardDrummer) [inline Test/tutorial/Simple_compiler/Compiler.dfy:270]: I don't see `Aggregate` or `FoldL` anywhere, although I do see `FoldR`

File: Test/tutorial/Simple_compiler/Compiler.dfy [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+
+  function method prettyPrint(p: Prog) : List<string> {
+    match p {
+      case Nil => Nil
+      case Cons(instr, p) => Cons(prettyPrintInstr(instr), prettyPrint(p))
+    }
+  }
+}
+
+/// 5. Exposing Dafny to C#
```
Comment (keyboardDrummer) [inline Test/tutorial/Simple_compiler/Compiler.dfy:442]: This one starts with 5. but the others are not numbered.

File: Test/tutorial/Simple_compiler/Compiler.dfy [hunk 6]
```diff
@@ -0,0 +1,498 @@
...
+  import StackMachine
+  import Translator
+  import Compiler
+  import PrettyPrint
+  import CSharpUtils
+  import Generics = System.Collections.Generic
+
+  class DafnyCompiler {
+    static method Compile(dAST: DafnyAST.Stmt) returns (dSM: StackMachine.Prog)
+      ensures DafnyAST.interpStmt(dAST) == StackMachine.interpProg(dSM)
```
Comment (keyboardDrummer) [inline Test/tutorial/Simple_compiler/Compiler.dfy:463]: Cool postcondition!! Would it be more a more complete example if these programs could also handle input? I was looking for the `forall input ::` but I see it doesn't apply to this language.

**Review by keyboardDrummer — APPROVED** [review]
Great example! Very nice :-)

File: Source/IntegrationTests/LitTests.cs [hunk 7]
```diff
@@ -104,7 +107,7 @@ public class LitTests {
           Environment.GetEnvironmentVariable("HOMEDRIVE") + Environment.GetEnvironmentVariable("HOMEPATH"));
         passthroughEnvironmentVariables = passthroughEnvironmentVariables
           .Concat(new[] {
-            "DOTNET_CLI_HOME",
+              "DOTNET_CLI_HOME",
```
Comment (MikaelMayer) [inline Source/IntegrationTests/LitTests.cs:110]: Please Remove extra space? Verify that there are no tabs as well.

File: Test/tutorial/Simple_compiler/README.md [hunk 8]
```diff
@@ -0,0 +1,3 @@
+# Using Dafny from C\#
+
+This directory shows how to call Dafny from C# (the reverse is possible as well and works similarly).  To make the example interesting we implement a small verified compiler from an arithmetic language to a stack machine.  The frontend is implemented in C# in `Main.cs`, and the backend is implemented in Dafny in `Compiler.dfy`.  The whole process
```
Comment (MikaelMayer) [inline Test/tutorial/Simple_compiler/README.md:3]: The whole process is great, can you finish your sentence?

**Review by MikaelMayer — COMMENTED** [review]
Looks great, I can't wait to have Dafny running Dafny !

File: Test/tutorial/Simple_compiler/Compiler.dfy (hunk 4 above)
Comment (cpitclaudel) [reply Test/tutorial/Simple_compiler/Compiler.dfy:270]: Good catch, thanks!

File: Test/tutorial/Simple_compiler/Compiler.dfy (hunk 3 above)
Comment (cpitclaudel) [reply Test/tutorial/Simple_compiler/Compiler.dfy:387]: Good point! I added more info with pointers to the following code

File: Source/IntegrationTests/LitTests.cs (hunk 7 above)
Comment (cpitclaudel) [reply Source/IntegrationTests/LitTests.cs:110]: Woops, thanks!

File: Test/tutorial/Simple_compiler/README.md (hunk 8 above)
Comment (cpitclaudel) [reply Test/tutorial/Simple_compiler/README.md]: !

**Comment by cpitclaudel** [comment]
Thanks @MikaelMayer and @keyboardDrummer for the reviews!  @keyboardDrummer , I added an additional constructor for inputs.  The commit that implements that is https://github.com/dafny-lang/dafny/pull/1769/commits/4ea62c8e2ae8ba8c7b96dfb13f7ac7b0a43ace61 .  Could you have a look and tell me whether you think the added complexity is worth it?  Personally I like it, I think it makes the example a bit more realistic!  

The theorem you were expecting is in a separate commit at https://github.com/dafny-lang/dafny/pull/1769/commits/2272b163831f987bcf37dbcf506528adfcab9d62

File: Test/tutorial/Simple_compiler/Compiler.dfy (hunk 6 above)
Comment (cpitclaudel) [reply Test/tutorial/Simple_compiler/Compiler.dfy:634]: Good point, I revamped the example to allow variables, so you can simulate input by running the program in a nonempty context.  Input in the sense of I/O would be harder to model (it requires nondeterminism)

Comment (keyboardDrummer) [reply Test/tutorial/Simple_compiler/Compiler.dfy:634]: Could you still change `CompileCorrect` to

```
  lemma compileCorrect(s: DafnyAST.Stmt)
    ensures forall input: DafnyAST.Context :: interpProg(compileStmt(s), input) == DafnyAST.interpStmt(s, input)
```

?

File: Test/tutorial/Simple_compiler/Compiler.dfy [hunk 9]
```diff
@@ -0,0 +1,671 @@
...
+    | Skip
+    | Print(e: Expr)
+    | Assign(v: string, e: Expr)
+    | Seq(s1: Stmt, s2: Stmt)
+
+/// The semantics of expressions and statements can be defined using recursive interpreters:
+
+  type Context = map<string, int>
+
+  function method interpExpr(e: Expr, ctx: Context): int {
```
Comment (keyboardDrummer) [inline Test/tutorial/Simple_compiler/Compiler.dfy:43]: Is `interpret` shortened to `interp` so it's not longer than `compile` ?

Comment (cpitclaudel) [reply Test/tutorial/Simple_compiler/Compiler.dfy:43]: It's a pretty common abbreviation, but I'm not sure why it's used ^^  Probably length, indeed?

File: Test/tutorial/Simple_compiler/Compiler.dfy (hunk 6 above)
Comment (cpitclaudel) [reply Test/tutorial/Simple_compiler/Compiler.dfy]: Sure thing, though of course that requires changing the types of the methods on both sides; done in the latest commit

File: Test/tutorial/Simple_compiler/Compiler.dfy [hunk 10]
```diff
@@ -0,0 +1,667 @@
...
+      case Skip =>
+        Skip
+      case Print(e) =>
+        Print(simplifyExpr(e))
+      case Assign(v, e) =>
+        Assign(v, simplifyExpr(e))
+      case Seq(s1, s2) =>
+        match (simplifyStmt(s1), simplifyStmt(s2)) {
+          case (s1', Skip) =>
+            assert forall o: seq<int> :: o + [] == o; // FIXME
```
Comment (keyboardDrummer) [inline Test/tutorial/Simple_compiler/Compiler.dfy:117]: FIXME

File: Test/tutorial/Simple_compiler/Compiler.dfy [hunk 11]
```diff
@@ -0,0 +1,667 @@
...
+            //           var (ctx1, o1) := interpStmt'(s1', ctx);
+            //           (ctx1, o1);
+            //         ==
+            //           interpStmt'(s1', ctx);
+            //       }
+            //     }
+            //   }
+            s1'
+          case (Skip, s2') =>
+            assert forall o: seq<int> :: [] + o == o; // WISH: why is this needed, why didn't I need it in the previous version (without contexts), and how should I have found this faster?
```
Comment (keyboardDrummer) [inline Test/tutorial/Simple_compiler/Compiler.dfy:153]: WISH?

**Review by keyboardDrummer — COMMENTED** [review]
Apart from the FIXME and WISH, looks good to me.

File: Test/tutorial/Simple_compiler/Compiler.dfy (hunk 6 above)
Comment (keyboardDrummer) [reply Test/tutorial/Simple_compiler/Compiler.dfy]: Thanks :-)

**Review by keyboardDrummer — APPROVED** [review]

**Pull request merged**